stan/math/rev/fun/tcrossprod.hpp

```diff
@@ -37,18 +37,8 @@ tcrossprod(const T& M) {
   Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime> MMt(M.rows(),
                                                                      M.rows());
 
-  vari** vs
-      = reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
-          (M.rows() * M.cols()) * sizeof(vari*)));
-  int pos = 0;
   for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < M.cols(); ++n) {
-      vs[pos++] = M(m, n).vi_;
-    }
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    MMt.coeffRef(m, m)
-        = var(new internal::dot_self_vari(vs + m * M.cols(), M.cols()));
+    MMt.coeffRef(m, m) = dot_self(M.row(m));
```

Comment:
I'm a bit concerned about these functions that use dot_product internally. I think we should just rewrite them to not lean on other functions.

I think with the rewrite we're saving to the autodiff stack O(N^3) varis when really we only need to save O(N^2) (if the input is an N*N matrix).

Reply:
> I think with the rewrite we're saving to the autodiff stack O(N^3) varis when really we only need to save O(N^2) (if the input is an N*N matrix).

That is a good point. Although I think we can still use prim dot_product.

Reply:
I will try to remember to just rewrite the autodiff for these. It shouldn't be hard to just do the whole thing.